### PR TITLE
Let Inspector edit a form submission

### PR DESCRIPTION
**Pull request by TimCsaky**
# Description

Allow an Inspector to edit 'location details' of a form submission.
(on the admin > submission review page, see 'edit submission' button)

this also addresses ticket [1112](https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-1112)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: app/frontend/src/components/admin/inspection/LocationEdit.vue [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+import Vuex from 'vuex';
+import { mapGetters } from 'vuex';
+
+import Location from '@/components/form/Location.vue';
+import { SilvipcRoles } from '@/utils/constants';
+
+/* eslint-disable */
+import ipcService from '../../../services/ipcService';
+
+const localStore = new Vuex.Store({
```
Comment (TimCsaky) [inline app/frontend/src/components/admin/inspection/LocationEdit.vue:42]: so.. is this where I create a 'local store'?
i'm not sure why i would lazy load anything if i have a store for this component

File: app/frontend/src/components/admin/inspection/LocationEdit.vue [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    };
+  },
+  computed: {
+    ...mapGetters('form', ['business', 'ipcPlan', 'location', 'submissionComplete']),
+    ...mapGetters('auth', ['hasSilvipcRoles']),
+
+    // Business data
+    businessName: {
+      get() { return this.business.name; },
+    },
```
Comment (TimCsaky) [inline app/frontend/src/components/admin/inspection/LocationEdit.vue:97]: i'm still populating the update form with data from the form store,
but i've taken out all the setters.

File: app/frontend/src/components/admin/inspection/LocationEdit.vue [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      // TODO reload location details in step2 component in case user changed details but didnt submit update
+      this.dialog = false;
+    },
+
+  },
+  mounted() {
+    // show edit button for inspectors on submission review page
+    this.showEditBtn = (this.$route.name == 'Submission' && this.hasSilvipcRoles([SilvipcRoles.INSPECTOR])) ? true : false;
+  },
+  created() {
```
Comment (TimCsaky) [inline app/frontend/src/components/admin/inspection/LocationEdit.vue:173]: not sure why i need to lazy load a store if it's already defined in this component.

File: app/frontend/src/components/admin/inspection/LocationEdit.vue (hunk 1 above)
Comment (jujaga) [reply app/frontend/src/components/admin/inspection/LocationEdit.vue:42]: You don't want to be defining a store at all here. When you lazy load a vuex module, the framework will bring that into memory.

File: app/frontend/src/components/admin/inspection/LocationEdit.vue (hunk 2 above)
Comment (jujaga) [reply app/frontend/src/components/admin/inspection/LocationEdit.vue:97]: Following the previous comment, you want the following Vue options in this module:
```js
  beforeDestroy() {
    this.$store.unregisterModule('edit');
  },
  created() {
    if(this.$store.hasModule('edit')) {
      this.$store.unregisterModule('edit');
    }
    this.$store.registerModule('edit', form);
  }
```
This would be how you lazy-load the form module into vuex, while putting it in a completely different namespace than 'form' to avoid editing the 'form' namespace state.

File: app/frontend/src/components/admin/inspection/LocationEdit.vue (hunk 3 above)
Comment (jujaga) [reply app/frontend/src/components/admin/inspection/LocationEdit.vue:173]: You want to lazy-load and unload a store because this allows you to control _when_ a namespaced module store is effective and mounted or not based on the visibility and lifecycle of a component (or set of them). We want to bind the lifecycle of your 'localStore" module to just this component, but also be able to continue to use the business logic in that module.

File: app/frontend/src/components/form/Location.vue [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+      // Todo, put in some utility fxn somewhere if needed again
+      numberOfWorkersRules: [
+        v => new RegExp('^[-+]?\\d+$').test(v) || 'invalid # of workers',
+        v => v > 0 || '# of workers must be greater than 0',
+        v => v < 9999 || '# of workers must 9999 or less'
+      ]
+    };
+  },
+  computed: {
+    ...mapGetters('form', ['location']),
```
Comment (jujaga) [inline app/frontend/src/components/form/Location.vue:229]: The one concern I have now that I think about this, is if we're using this Location component in multiple places, but it's pulling from not necessarily the right store. This could introduce a subtle and hard to trace bug - would need to double check that behaviors on all pathways are doing what you think it should be doing.

File: app/frontend/src/components/form/Location.vue [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+    ...mapGetters('auth', ['hasSilvipcRoles']),
+
+    // Location
+    startDate: {
+      get() {
+        return this.location.startDate;
+      },
+      set(value) {
+        this.updateLocation({ ['startDate']: value });
+      }
```
Comment (jujaga) [inline app/frontend/src/components/form/Location.vue:239]: Style: These should be written as one-liners. Don't need to expand the get/set functions as they're just mapping and doing one thing.

File: app/frontend/src/router/index.js [hunk 6]
```diff
@@ -62,7 +63,16 @@ const router = new VueRouter({
...
-  ]
+  ],
+
+  scrollBehavior: function (to, from, savedPosition) { // eslint-disable-line
+    if (to.hash) {
+      return { selector: to.hash };
+    } else {
+      return { x: 0, y: 0 };
+    }
+  },
```
Comment (jujaga) [inline app/frontend/src/router/index.js:74]: What is this function for? Is this necessary?

File: app/frontend/src/store/modules/form.js [hunk 7]
```diff
@@ -212,6 +212,7 @@ export default {
       } finally {
         commit('setSubmitting', false);
       }
-    }
+    },
```
Comment (jujaga) [inline app/frontend/src/store/modules/form.js:215]: Unnecessary trailing comma

File: app/frontend/src/components/admin/AdminReviewSubmission.vue [hunk 8]
```diff
@@ -45,7 +45,6 @@ import Step2 from '@/components/form/Step2.vue';
 import Step3 from '@/components/form/Step3.vue';
 import Step4 from '@/components/form/Step4.vue';
 import Step5 from '@/components/form/Step5.vue';
-
```
Comment (jujaga) [inline app/frontend/src/components/admin/AdminReviewSubmission.vue:48]: Not sure why this whitespace line was removed.

File: app/frontend/src/components/admin/inspection/LocationEdit.vue [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+      dialog: false,
+      updateValid: false,
+      reviewMode: false,
+      startDateMenu: false,
+      endDateMenu: false,
+    };
+  },
+  computed: {
+    ...mapGetters('form', ['business', 'ipcPlan', 'location', 'submissionComplete']),
+    ...mapActions('form', ['updateForm']),
```
Comment (TimCsaky) [inline app/frontend/src/components/admin/inspection/LocationEdit.vue:61]: so if we have a 'copy' of the form store in a new 'edit' namespace..
i assumed this line shoul dread:
`...mapActions('edit', ['updateForm']),`
but it doesnt find that function.

**Comment by TimCsaky** [comment]
for some reason the city look up field isnt populated in the edit form.. or the main form when a user goes back to update

File: app/frontend/src/components/admin/inspection/LocationEdit.vue [hunk 10]
```diff
@@ -0,0 +1,140 @@
+<template>
+  <v-dialog v-model="dialog" max-width="800px">
```
Comment (jujaga) [inline app/frontend/src/components/admin/inspection/LocationEdit.vue:2]: Don't we have the BaseDialog.vue component in silvipc? We should be leveraging that component if possible.

File: app/frontend/src/components/admin/inspection/LocationEdit.vue [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+  //   // dynamically register a ephemeral vuex module for this 'edit location' component
+  //   if(this.$store.hasModule('edit')) {
+  //     this.$store.unregisterModule('edit');
+  //   }
+  //   this.$store.registerModule('edit', form);
+  // },
+  // beforeDestroy() {
+  //   // unload this store
+  //   this.$store.unregisterModule('edit');
+  // },
```
Comment (jujaga) [inline app/frontend/src/components/admin/inspection/LocationEdit.vue:125]: If you mount the form store in the namespace "edit", your mapgetter needs to pull from the "edit" namespace, not "form".

File: app/frontend/src/components/admin/inspection/LocationEdit.vue [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+      updateValid: false,
+      reviewMode: false,
+      updateComplete: false,
+      updateError: '',
+      startDateMenu: false,
+      endDateMenu: false
+    };
+  },
+  computed: {
+    ...mapGetters('form', ['business', 'contacts', 'covidContact', 'ipcPlan', 'location']),
```
Comment (jujaga) [inline app/frontend/src/components/admin/inspection/LocationEdit.vue:61]: If you mount the form store in the namespace "edit", your mapgetter needs to pull from the "edit" namespace, not "form". This referencing form instead is why it doesn't look like lazy loading is working.

File: app/frontend/src/components/admin/inspection/LocationEdit.vue [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+          throw new Error('No response data from API while submitting update');
+        }
+        this.updateComplete = true;
+      } catch (error) {
+        console.error(`Error submitting update: ${error}`); // eslint-disable-line no-console
+        this.updateError = 'An error occurred while attempting to update the form. Please try again.';
+      }
+    },
+
+    async cancelUpdate(){
```
Comment (jujaga) [inline app/frontend/src/components/admin/inspection/LocationEdit.vue:106]: No await used - doesn't need to be async function.

File: app/frontend/src/store/modules/form.js [hunk 14]
```diff
@@ -213,5 +213,6 @@ export default {
         commit('setSubmitting', false);
       }
     }
+
```
Comment (jujaga) [inline app/frontend/src/store/modules/form.js:216]: Unnecessary whitespace line added

File: app/frontend/src/views/admin/Submission.vue [hunk 15]
```diff
@@ -87,6 +95,8 @@ export default {
       this.$refs.notesPanel.getNotes();
     }
   },
+  mounted() {
+  },
```
Comment (jujaga) [inline app/frontend/src/views/admin/Submission.vue:99]: Don't need a mounted here if it's doing nothing.

File: app/frontend/src/components/admin/inspection/LocationEdit.vue (hunk 10 above)
Comment (TimCsaky) [reply app/frontend/src/components/admin/inspection/LocationEdit.vue:2]: it doesnt really work for a form

File: app/frontend/src/components/admin/inspection/LocationEdit.vue (hunk 13 above)
Comment (TimCsaky) [reply app/frontend/src/components/admin/inspection/LocationEdit.vue]: good point

File: app/frontend/src/components/admin/inspection/LocationEdit.vue [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+    cancelUpdate(){
+      this.dialog = false;
+      // reload form data in parent component - from the global 'from' state module
+      this.getForm(this.ipcPlanId);
+    },
+
+  },
+  // lazy-load a local copy of the 'form' store into the 'edit' namespace
+  beforeCreate() {
```
Comment (TimCsaky) [inline app/frontend/src/components/admin/inspection/LocationEdit.vue:135]: needed to use beforeCreate() instead.
So i can now use the form module in a 'local' namespace but i havent found a way to override the getters and setters in the child Location component.

File: app/frontend/src/components/form/Location.vue [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+          </template>
+          <v-date-picker v-model="endDate" @input="endDateMenu = false" :readonly="reviewMode"></v-date-picker>
+        </v-menu>
+      </v-col>
+    </v-row>
+
+    <v-row>
+      <v-col cols="12" sm="6" lg="5">
+        <label>Closest Community / Town / City</label>
+        <CityLookup
```
Comment (TimCsaky) [inline app/frontend/src/components/form/Location.vue:72]: having trouble getting this field to populate in admin > review and edit form

**Comment by TimCsaky** [comment]
i wasnt able to pass the store name as a property to the child Location component.
i tried some of the things here but had no luck:
https://stackoverflow.com/questions/42585055/pass-prop-as-module-name-when-mapping-to-namespaced-module

I'm thinking this will work with the single store. and frankly, i cant waste another day trying to get it working.

now.. the only issue i see is that teh city lookup combobox isnt populated. i'm not sure if i need another listener in there or what. i tried a few things but had no luck. I'd love to know if you see anything obvious I'm missing.

i've been going a little crazy with this ticket. learnt some vue stuff.. but it sure is dragging on.

**Comment by TimCsaky** [comment]
oh. i've just noticed i left the city related fields visible on all pages to try and debug the issue.

**Comment by TimCsaky** [comment]
also looks like the sort order for the submissions table is not working again

File: app/frontend/src/components/bcgov/BCGovHeader.vue [hunk 18]
```diff
@@ -11,20 +11,43 @@
...
+    },
+    // clicking on banner logo or title goes to /home or /admin if an admin user
+    followBannerLink() {
+      if(this.hasRole(this.role.INSPECTOR) & !this.hasRole(this.role.DEVELOPER)){
+        this.$router.push({ name: 'Admin' }).catch(() => {});
+      }
+      else{
+        this.$router.push({ name: 'Home' }).catch(() => {});
+      }
+    }
```
[the remark this answers is not in the document]
Comment (TimCsaky) [reply app/frontend/src/components/bcgov/BCGovHeader.vue]: sure.. things can be more compact.. i've found a compromise

**Comment by loneil** [comment]
I see the sort order working, and the city lookup working on this PR. What's the problem with those?

**Comment by TimCsaky** [comment]
> 
> 
> I see the sort order working, and the city lookup working on this PR. What's the problem with those?

for me (both on my local and in the pr, both rebased with dev master), i see the submission table sorted created ASC (latest submission in the last row).
when an inspector goes to the edit location modal, the city look up field doesnt show the city.

on a side note.. in a previous commit i made my code for the home (logo and tile) links in the banner a bit more succinct using <view-router> element. this broke the BCGovHeader unit test so i reverted that. not sure it's worth spending more time on that.

File: app/frontend/src/components/form/CityLookup.vue [hunk 19]
```diff
@@ -26,15 +26,19 @@ import Vue from 'vue';
...
+    fieldModel:String,
     fieldRules: Array,
   },
   data() {
     return {
       isLoading: false,
       features: [],
       search: null,
-      model: this.fieldModel,
+      model: (this.fieldModel) ? this.fieldModel : this.currentValue,
```
Comment (jujaga) [inline app/frontend/src/components/form/CityLookup.vue:41]: Don't need parentheses for the conditional.
